src/runtime/task/mod.rs

```diff
@@ -285,3 +324,55 @@ impl Debug for TaskSet {
         write!(f, "}}")
     }
 }
+
+/// A unique identifier for a [`LocalKey`](crate::thread::LocalKey)
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+struct LocalKeyId(usize);
+
+impl LocalKeyId {
+    fn new<T: 'static>(key: &'static LocalKey<T>) -> Self {
+        Self(key as *const _ as usize)
+    }
+}
+
+/// A map of thread-local storage values. Values are Option<_> because we need to be able to
+/// iteratively destruct them, as it's valid for TLS destructors to initialize new TLS slots.
+struct LocalMap(HashMap<LocalKeyId, Option<Box<dyn Any>>>);
+
+impl LocalMap {
+    fn new() -> Self {
+        Self(HashMap::new())
+    }
+
+    fn get<T: 'static>(&self, key: &'static LocalKey<T>) -> Option<Result<&T, AlreadyDestructedError>> {
+        self.0.get(&LocalKeyId::new(key)).map(|val| {
+            val.as_ref()
+                .map(|val| {
+                    Ok(val
+                        .downcast_ref::<T>()
+                        .expect("local value must downcast to expected type"))
+                })
+                .unwrap_or(Err(AlreadyDestructedError))
+        })
+    }
+
+    fn init<T: 'static>(&mut self, key: &'static LocalKey<T>, value: T) {
+        let key = LocalKeyId::new(key);
+        assert!(self.0.get(&key).is_none(), "cannot reinitialize a TLS slot");
```

Comment:
Instead of doing a `get` followed by an `insert`, use the `entry()` api instead?